### PR TITLE
xmm_newton: Adding epic images, epic spectra and epic light curves

### PR DESCRIPTION
**Pull request by esdc-esac-esa-int**

### CODE REVIEW

File: astroquery/esa/xmm_newton/core.py [hunk 1]
```diff
@@ -297,9 +301,146 @@ def _parse_filename(self, filename):
...
+                        key = fname_info["I"] + "_bkg"
+                    elif fname_info["T"] == "SRCARF":
+                        key = fname_info["I"] + "_arf"
+                    else:
+                        with fits.open(value) as hdul:
+                            for ext in hdul:
+                                if ext.name != "SPECTRUM":
+                                    continue
+                                rmf_fname = ext.header["RESPFILE"]
+                                if fname_info["I"] == "M1" or \
```
Comment (bsipocz) [inline astroquery/esa/xmm_newton/core.py:391]: no need for line breaks, and below, we let longer than 80char long lines, the limit in the linter is set to 120, so keep the logical blocks together if they fit within that limit.

Removing the breaks would improve overall readability.

File: astroquery/esa/xmm_newton/core.py [hunk 2]
```diff
@@ -297,9 +301,146 @@ def _parse_filename(self, filename):
...
+                                if verbose:
+                                    log.info("rmf link is: %s" % link)
+
+                                response = self._request('GET', link)
+
+                                rsp_filename =\
+                                    os.path.join(_path,
+                                                 paths[0],
+                                                 paths[1],
+                                                 ext.header["RESPFILE"])
```
Comment (bsipocz) [inline astroquery/esa/xmm_newton/core.py:413]: These all should be in one line, I very much find it difficult to read this one word in a line formatting while the logical block is one single expression

File: astroquery/esa/xmm_newton/tests/test_xmm_newton.py [hunk 3]
```diff
@@ -64,6 +68,15 @@ def test_get_postcard(self):
         xsa = XMMNewtonClass(self.get_dummy_tap_handler())
         xsa.get_postcard(**parameters)
 
+    @pytest.mark.remote_data
```
Comment (bsipocz) [inline astroquery/esa/xmm_newton/tests/test_xmm_newton.py:71]: This test file gets nicely long, maybe it's time to factor out the remote tests to a separate `test_xmm_newton_remote.py`. (See most other modules have a `test_modulename.py` and a `test_modulename_remote.py`, it makes it's easier to spot issues. Also, when it makes sense tests can be duplicated between the two, e.g. we always run the mocked local ones, but only cron/merge run the remote ones. The previous tests that the code internally consistent, but it doesn't provide any safe guarding against remote API changes).

File: astroquery/esa/xmm_newton/tests/test_xmm_newton.py [hunk 4]
```diff
@@ -73,15 +86,20 @@ def test_query_xsa_tap(self):
...
     def test_get_tables(self):
         parameters2 = {'only_names': True,
                        'verbose': True}
 
-        dummyTapHandler = DummyXMMNewtonTapHandler("get_tables", parameters2)
-        xsa = XMMNewtonClass(self.get_dummy_tap_handler())
-        xsa.get_tables(only_names=True, verbose=True)
-        dummyTapHandler.check_call("get_tables", parameters2)
+        dummyHandler = DummyXMMNewtonTapHandler("get_tables", parameters2)
+        dummyHandler.check_call("get_tables", parameters2)
```
Comment (bsipocz) [inline astroquery/esa/xmm_newton/tests/test_xmm_newton.py:100]: The previous approach was more preferred as it calls the user facing API. As I see (I may be mistaken), here there isn't an  instance of `XMMNewtonClass` called. This infact also the case in a few other tests.

File: astroquery/esa/xmm_newton/tests/test_xmm_newton.py [hunk 5]
```diff
@@ -171,14 +283,140 @@ def _create_tar(self, tarname, files):
...
+        _source_number = 83
+        _instruments = ["M1", "M1_arf", "M1_bkg", "M1_rmf",
+                        "M2", "M2_arf", "M2_bkg", "M2_rmf",
+                        "PN", "PN_arf", "PN_bkg", "PN_rmf"]
+        self._create_tar(_tarname, self._files)
+        xsa = XMMNewtonClass(self.get_dummy_tap_handler())
+        res = xsa.get_epic_spectra(_tarname, _source_number,
+                                   instrument=[])
+        assert len(res) == 8
+        '''
```
Comment (bsipocz) [inline astroquery/esa/xmm_newton/tests/test_xmm_newton.py:382]: Remove these commented out lines

**Review by bsipocz — COMMENTED** [review]
This will need a changelog, too. 

Some squashing would be nice for the first 3 commits (e.g. there is the unrelated `get_product_from_4xmm_catalogue` method added and then reverted), but it's not critical.

I would also ask @ceb8 or @keflavich to have a second look at the tests.

File: astroquery/esa/xmm_newton/tests/test_xmm_newton.py (hunk 3 above)
Comment (keflavich) [reply astroquery/esa/xmm_newton/tests/test_xmm_newton.py:71]: on this line also - this is marked as "remote data", but it uses "dummy_tap_handler" - doesn't that mean it is not doing a remote request?

File: astroquery/esa/xmm_newton/core.py [hunk 6]
```diff
@@ -406,31 +407,25 @@ def get_epic_spectra(self, filename, source_number, *,
...
-                        ret[key].append(value)
+                        ret[key].append(path_inst_name)
                     else:
-                        ret[key] = value
+                        ret[key] = path_inst_name
 
         except FileNotFoundError:
             log.error("File %s not found" % (filename))
-            return {}
+            return None
```
Comment (keflavich) [inline astroquery/esa/xmm_newton/core.py:426]: `return None` is redundant; this can just be `return`

File: astroquery/esa/xmm_newton/core.py [hunk 7]
```diff
@@ -406,31 +407,25 @@ def get_epic_spectra(self, filename, source_number, *,
...
 
-        if ret == {}:
+        if ret is None or ret == {}:
```
Comment (keflavich) [inline astroquery/esa/xmm_newton/core.py:428]: this is fine, but, please verify that `if not ret:` is not a valid equivalent here

File: astroquery/esa/xmm_newton/core.py [hunk 8]
```diff
@@ -97,6 +99,8 @@ def download_data(self, observation_id, *, filename=None, verbose=False,
         None if not verbose. It downloads the observation indicated
         If verbose returns the filename
         """
+        if filename is not None:
+            filename = filename.split('.')[0]
```
[the remark this answers is not in the document]
Comment (esdc-esac-esa-int) [reply astroquery/esa/xmm_newton/core.py]: Done, thanks

File: astroquery/esa/xmm_newton/core.py [hunk 9]
```diff
@@ -297,9 +301,146 @@ def _parse_filename(self, filename):
...
+        if ret == {}:
+            log.info("Nothing to extract with the given parameters:\n"
+                     "  PPS: %s\n"
+                     "  Source Number: %u\n"
+                     "  Instrument: %s\n" % (filename, source_number,
+                                             instrument))
+
+        return ret
+
+    def get_epic_images(self, filename, band=[], instrument=[], **kwargs):
```
[the remark this answers is not in the document]
Comment (esdc-esac-esa-int) [reply astroquery/esa/xmm_newton/core.py]: Done, thanks

File: astroquery/esa/xmm_newton/core.py [hunk 10]
```diff
@@ -297,9 +301,146 @@ def _parse_filename(self, filename):
...
+        If the instrument is not specified this function will
+        return all the available instruments
+        Examples:
+        Extracting all bands and instruments:
+            result = XMM.get_epic_spectra(tarfile,83,
+                                         instrument=['M1','M2','PN'])
+        If we want to retrieve the source spectrum of the instrument PN
+            fits_image = result['PN']
+        fits_image will be the full path to the extracted FTZ file
+        Parameters
```
[the remark this answers is not in the document]
Comment (esdc-esac-esa-int) [reply astroquery/esa/xmm_newton/core.py:325]: Done, thanks

File: astroquery/esa/xmm_newton/core.py [hunk 11]
```diff
@@ -297,9 +301,146 @@ def _parse_filename(self, filename):
...
+        """Extracts the EPIC sources spectral products from a given TAR file
+        For a given TAR file obtained with:
+            XMM.download_data(OBS_ID,level="PPS",extension="FTZ",filename=tarfile)
+        This function extracts the EPIC sources spectral products in a given
+        instrument (or instruments) from it
+        The result is a dictionary containing the paths to the extracted EPIC
+        sources spectral products with key being the instrument
+        If the instrument is not specified this function will
+        return all the available instruments
+        Examples:
```
[the remark this answers is not in the document]
Comment (esdc-esac-esa-int) [reply astroquery/esa/xmm_newton/core.py:317]: Done, thanks

File: astroquery/esa/xmm_newton/core.py [hunk 12]
```diff
@@ -297,9 +301,146 @@ def _parse_filename(self, filename):
...
+        _product_type = ["SRSPEC", "BGSPEC", "SRCARF"]
+        _path = ""
+        ret = {}
+        if instrument == []:
+            instrument = _instrument
+        else:
+            for i in instrument:
+                if i not in _instrument:
+                    log.warning("Invalid instrument %s" % i)
+                    instrument.remove(i)
```
[the remark this answers is not in the document]
Comment (esdc-esac-esa-int) [reply astroquery/esa/xmm_newton/core.py]: Done, thanks

File: astroquery/esa/xmm_newton/core.py [hunk 13]
```diff
@@ -297,9 +301,146 @@ def _parse_filename(self, filename):
...
+        if path != "" and os.path.exists(path):
+            _path = path
+        try:
+            with tarfile.open(filename, "r") as tar:
+                for i in tar.getmembers():
```
[the remark this answers is not in the document]
Comment (esdc-esac-esa-int) [reply astroquery/esa/xmm_newton/core.py]: Done, thanks

File: astroquery/esa/xmm_newton/core.py [hunk 14]
```diff
@@ -297,9 +301,146 @@ def _parse_filename(self, filename):
...
+                        continue
+                    if not fname_info["I"] in instrument:
+                        continue
+                    if not fname_info["T"] in _product_type:
+                        continue
+                    if int(fname_info["X-"], 16) != source_number:
+                        continue
+                    tar.extract(i, _path)
+                    key = fname_info["I"]
+                    value = os.path.abspath(os.path.join(_path, i.name))
```
[the remark this answers is not in the document]
Comment (esdc-esac-esa-int) [reply astroquery/esa/xmm_newton/core.py]: Done, thanks

File: astroquery/esa/xmm_newton/core.py (hunk 1 above)
Comment (esdc-esac-esa-int) [reply astroquery/esa/xmm_newton/core.py]: Done, thanks

File: astroquery/esa/xmm_newton/core.py (hunk 2 above)
Comment (esdc-esac-esa-int) [reply astroquery/esa/xmm_newton/core.py]: Done, thanks

File: astroquery/esa/xmm_newton/tests/test_xmm_newton.py (hunk 4 above)
Comment (esdc-esac-esa-int) [reply astroquery/esa/xmm_newton/tests/test_xmm_newton.py]: Done, thanks

File: astroquery/esa/xmm_newton/core.py (hunk 7 above)
Comment (esdc-esac-esa-int) [reply astroquery/esa/xmm_newton/core.py]: it is indeed equivalent. Changed. Thanks

File: astroquery/esa/xmm_newton/core.py (hunk 6 above)
Comment (esdc-esac-esa-int) [reply astroquery/esa/xmm_newton/core.py]: Changed, thanks

File: astroquery/esa/xmm_newton/core.py [hunk 15]
```diff
@@ -297,9 +301,146 @@ def _parse_filename(self, filename):
...
+        _instrument = ["M1", "M2", "PN", "EP"]
+        _product_type = ["SRSPEC", "BGSPEC", "SRCARF"]
+        _path = ""
+        ret = {}
+        if instrument == []:
+            instrument = _instrument
+        else:
+            for i in instrument:
+                if i not in _instrument:
+                    log.warning("Invalid instrument %s" % i)
```
[the remark this answers is not in the document]
Comment (esdc-esac-esa-int) [reply astroquery/esa/xmm_newton/core.py]: Done, thanks

File: astroquery/esa/xmm_newton/core.py [hunk 16]
```diff
@@ -297,9 +301,146 @@ def _parse_filename(self, filename):
...
+                    instrument.remove(i)
+        if path != "" and os.path.exists(path):
+            _path = path
+        try:
+            with tarfile.open(filename, "r") as tar:
```
[the remark this answers is not in the document]
Comment (esdc-esac-esa-int) [reply astroquery/esa/xmm_newton/core.py:366]: filename is an input parameter that represents the name of the tarfile to be processed.
path, on the other hand, if set, represents the path where EPIC images will be extracted

File: astroquery/esa/xmm_newton/core.py [hunk 17]
```diff
@@ -297,9 +301,146 @@ def _parse_filename(self, filename):
...
+        as this is the convention used by the pipeline.
+        The structure and the content of the extracted compressed FITS files
+        are described in details in the Pipeline Products Description
+        [XMM-SOC-GEN-ICD-0024](https://xmm-tools.cosmos.esa.int/external/xmm_obs_info/odf/data/docs/XMM-SOC-GEN-ICD-0024.pdf).
+        """
+        _instrument = ["M1", "M2", "PN", "EP"]
+        _product_type = ["SRSPEC", "BGSPEC", "SRCARF"]
+        _path = ""
+        ret = {}
+        if instrument == []:
```
[the remark this answers is not in the document]
Comment (esdc-esac-esa-int) [reply astroquery/esa/xmm_newton/core.py:356]: Done, thanks

**Comment by esdc-esac-esa-int** [comment]
I think we went trough all of the issues raised. Please, let us know if it is ready to be merged, thanks!

**Comment by ceb8** [comment]
Thanks @esdc-esac-esa-int !

**Pull request merged**